clang/lib/CodeGen/CoverageMappingGen.cpp

```diff
@@ -2177,7 +2177,8 @@ struct CounterCoverageMappingBuilder
   }
 
   void VisitOpaqueValueExpr(const OpaqueValueExpr* OVE) {
-    Visit(OVE->getSourceExpr());
+    if (const Expr *SE = OVE->getSourceExpr())
```

Comment:
I suspect the correct check here is `if (OVE->isUnique())`?  See https://reviews.llvm.org/D39562 .  CC @ahatanak.

Reply:
Not all `OpaqueValueExpr`s having a source expression are marked as "unique". Checking `isUnique()` instead of `getSourceExpr()` breaks at least handling of the GNU extension of the `?:` operator:
```
    1|      1|int main() {
    2|      1|    int i = 1;
    3|      1|    return (i ? 1 : 0)
                                  ^0        <-- disappears if isUnique() is used
  ------------------
  |  Branch (3:12): [True: 1, False: 0]
  |  Branch (3:13): [True: 1, False: 0]     <-- disappears if isUnique() is used
  ------------------
    4|      1|        ?: 10;
                         ^0
    5|      1|}
```

Reply:
If you have a BinaryConditionalOperator, I'm not sure this does the right thing.  If I'm following correctly, you end up visiting the condition twice, which is not accurately reflecting the way this actually executes.  The correct way to visit a BinaryConditionalOperator is to visit the "common" expression (BinaryConditionalOperator::getCommon()) first, then don't recurse into the OpaqueValueExpr.

Reply:
> If I'm following correctly, you end up visiting the condition twice

No, the "true" branch visitation (`propagateCounts`) is performed only when it is the ordinary `ConditionalOperator` (https://github.com/llvm/llvm-project/blob/main/clang/lib/CodeGen/CoverageMappingGen.cpp#L2018).

Do you think it should be rewritten first? It's just the first case found by me where a non-unique OVE has a source expression which is visited. There may be more cases.

Reply:
The point of the "unique" flag is precisely to indicate whether a subexpression is naturally part of the tree at that point, or it's referring to an expression which should have been evaluated elsewhere.  So not visiting a non-unique expression should be the right default.

I didn't realize the existing AbstractConditionalOperator code already had a workaround for this... we should be able to adjust it to compensate.

Reply:
Makes sense, thanks! Moreover, I've really found a couple of bugs when some expressions referred by `OpaqueValueExpr` are handled more than once (#88881 and #88898). Regarding `BinaryConditionalOperator`, I've decided to refactor it [in a separate PR](https://github.com/llvm/llvm-project/pull/88910) and then to introduce the `isUnique()` check in this PR, hoping that there will be no more regressions.